klippa/src/gdef.rs

```diff
@@ -1,11 +1,488 @@
 //! impl subset() for GDEF
 
-use crate::{CollectVariationIndices, Plan};
-use write_fonts::read::{
-    collections::IntSet,
-    tables::gdef::{CaretValue, Gdef, LigCaretList, LigGlyph, MarkGlyphSets},
-    types::GlyphId,
+use crate::{
+    layout::ClassDefSubsetStruct,
+    offset::{SerializeSerialize, SerializeSubset},
+    offset_array::SubsetOffsetArray,
+    serialize::{SerializeErrorFlags, Serializer},
+    CollectVariationIndices, Plan, Subset, SubsetError, SubsetTable,
 };
+use write_fonts::{
+    read::{
+        collections::IntSet,
+        tables::{
+            gdef::{
+                AttachList, AttachPoint, CaretValue, CaretValueFormat1, CaretValueFormat2,
+                CaretValueFormat3, Gdef, LigCaretList, LigGlyph, MarkGlyphSets,
+            },
+            layout::CoverageTable,
+        },
+        types::GlyphId,
+        FontRef, TopLevelTable,
+    },
+    types::Offset16,
+    FontBuilder,
+};
+
+// reference: subset() for GDEF in harfbuzz
+// <https://github.com/harfbuzz/harfbuzz/blob/59001aa9527c056ad08626cfec9a079b65d8aec8/src/OT/Layout/GDEF/GDEF.hh#L660>
+impl Subset for Gdef<'_> {
+    fn subset(
+        &self,
+        plan: &Plan,
+        _font: &FontRef,
+        s: &mut Serializer,
+        _builder: &mut FontBuilder,
+    ) -> Result<(), SubsetError> {
+        subset_gdef(self, plan, s).map_err(|_| SubsetError::SubsetTableError(Gdef::TAG))
+    }
+}
+
+fn subset_gdef(gdef: &Gdef, plan: &Plan, s: &mut Serializer) -> Result<(), SerializeErrorFlags> {
+    let version = gdef.version();
+    // major version
+    s.embed(version.major)?;
+
+    // minor version, might change after subset
+    let minor_version_pos = s.embed(version.minor)?;
+
+    // glyph classdef offset
+    let glyph_classdef_offset_pos = s.embed(0_u16)?;
+
+    // attach list offset
+    let attachlist_offset_pos = s.embed(0_u16)?;
+
+    // ligcaret list offset
+    let ligcaret_list_offset_pos = s.embed(0_u16)?;
+
+    //mark attach classdef offset
+    let markattach_classdef_offset_pos = s.embed(0_u16)?;
+
+    let snapshot_version0 = s.snapshot();
+
+    let mark_glyphsetsdef_offset_pos = if version.minor >= 2 {
+        s.embed(0_u16)?
+    } else {
+        0
+    };
+
+    let mut subset_varstore = false;
+    // TODO: make sure repacker will not move the target subtable before the other children
+    // ref: <https://github.com/harfbuzz/harfbuzz/blob/aad5780f5305f38ef128c61854c5d5a0c4ca3f4f/src/OT/Layout/GDEF/GDEF.hh#L665>
+    // Push var store first (if it's needed) so that it's last in the
+    // serialization order. Some font consumers assume that varstore runs to
+    // the end of the GDEF table.
+    // See: https://github.com/harfbuzz/harfbuzz/issues/4636
+
+    if version.minor >= 3 {
```

Comment:
Instead of having subset_varstore be mutable you could do this:

let subset_varstore = if version.minor >= 3 { ... } else { false }

Likewise for all of the other ones below as well.